### PR TITLE
fix(eslint-plugin): [switch-exhaustiveness-check] handle special characters in enum keys

### LINKED ISSUE #2181: [switch-exhaustiveness-check] Autocorrect incorrect parsing/handling property names with hyphens in them
**Repro**

Just have an enum with at least one property containing a hyphen:

```TS
export enum MyEnum {
	'blah-blah' = 'blah-blah',
	'test' = 'test'
}
```

```JSON
{
  "rules": {
    "@typescript-eslint/switch-exhaustiveness-check": "error"
  }
}
```

```TS
function myFunction(
		arg: MyEnum
	): string {
		switch (reason) {

		}
	}

```

**Expected Result**

```TS
function myFunction(
		arg: MyEnum
	): string {
		switch (reason) {
case MyEnum['blah-blah']: { throw new Error('Not implemented yet: MyEnum.['blah-blah'] case') }
case MyEnum.test: { throw new Error('Not implemented yet: MyEnum.test case') }		}
	}

```

**Actual Result**

```TS
function myFunction(
		arg: MyEnum
	): string {
		switch (reason) {
case MyEnum.blah-blah: { throw new Error('Not implemented yet: MyEnum.blah-blah case') }
case MyEnum.test: { throw new Error('Not implemented yet: MyEnum.test case') }		}
	}

```
**Additional Info**

As you can see the above code will error, as a property name containing hyphens hasn't been put in quotes.

On latest version of typescript@eslint-plugin 2.33.0

### PR DESCRIPTION
**Pull request by karishnu**
Fixes #2181

### CODE REVIEW

**Comment by karishnu** [comment]
@bradzacher Are there any changes required before this can be merged?

**Comment by karishnu** [comment]
@bradzacher Are any changes required before this can be merged?

**Comment by bradzacher** [comment]
This PR is in the queue of PRs to reviewed, and will be reviewed when we are able.
https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#addressing-feedback-and-beyond

File: packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts [hunk 1]
```diff
@@ -42,7 +42,9 @@ export default createRule({
       fixer: TSESLint.RuleFixer,
       node: TSESTree.SwitchStatement,
       missingBranchTypes: Array<ts.Type>,
+      symbolName?: string,
     ): TSESLint.RuleFix | null {
+      const identifierRegex = /^[a-zA-Z_$][0-9a-zA-Z_$]*$/;
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts:47]: The problem we've got is that `a-zA-Z` doesn't include all of the "letter-like" characters that are actually allowed in the chinese/japanese/korean/cyrillic/etc characters, accented characters, etc. So for a large number of codebases this regex will actually cause a false match. A false match isn't the end of the world, but it's going to be annoying for non-english speakers.

This is the actual regex of allowed characters within a variable name: https://github.com/estools/esutils/blob/a825f91fd1d1e3a9fff84227cb06c011d8a0b9e8/lib/code.js#L39-L44

It's a mess.

TypeScript needs to parse identifiers itself, and thankfully they export two functions that we can use here: `isIdentifierStart` and `isIdentifierPart`.

You can use this function instead to determine if the name requires quoting:
```ts
const compilerOptions = service.program.getCompilerOptions();
function requiresQuoting(name: string): boolean {
  if (name.length === 0) {
    return true;
  }
  if (!ts.isIdentifierStart(name.charCodeAt(0), compilerOptions.target)) {
    return true;
  }

  for (let i = 1; i < name.length; i += 1) {
    if (!ts.isIdentifierPart(name.charCodeAt(i), compilerOptions.target)) {
      return true;
    }
  }

  return false;
}
```

Comment (karishnu) [reply packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts:47]: Thank you! Will make the changes as soon as I can find some time.

**Pull request merged**